velox/functions/prestosql/aggregates/ReservoirSampleAggregate.cpp

```diff
@@ -0,0 +1,663 @@
+      processInitialSample(accumulator, initialSampleDecoded, row, elementType);
+    }
+  }
+
+  void processInitialSample(
+      ReservoirSampleAccumulator* accumulator,
+      const DecodedVector& initialSampleDecoded,
+      vector_size_t row,
+      const TypePtr& elementType) {
+    auto* arrayVector = initialSampleDecoded.base()->as<ArrayVector>();
+    auto arrayIndex = initialSampleDecoded.index(row);
+    auto arraySize = arrayVector->sizeAt(arrayIndex);
+    auto arrayOffset = arrayVector->offsetAt(arrayIndex);
+
+    if (arraySize > 0) {
+      VELOX_CHECK(
+          accumulator->initialSeenCount >= arraySize,
+          "initialProcessedCount must be greater than or equal to the number of positions in the initial sample");
+      auto elementVector = arrayVector->elements();
+      accumulator->initialSamples =
+          BaseVector::create(elementType, arraySize, allocator_->pool());
+      accumulator->initialSamples->resize(arraySize);
+      accumulator->initialSamples->copy(
+          elementVector.get(), 0, arrayOffset, arraySize);
+      accumulator->samples =
+          BaseVector::create(elementType, arraySize, allocator_->pool());
+      accumulator->samples->resize(arraySize);
+      accumulator->samples->copy(
+          elementVector.get(), 0, arrayOffset, arraySize);
+    }
+  }
+
+  void mergeIntermediateSamples(
+      ReservoirSampleAccumulator* accumulator,
+      const ArrayVector* sampleArrayVector,
+      const ArrayVector* initialSampleArrayVector,
+      const FlatVector<int64_t>* seenCountVector,
+      vector_size_t decodedRow,
+      const TypePtr& elementType) {
+    int64_t otherProcessedCount = seenCountVector->valueAt(decodedRow);
+    auto otherSampleSize = sampleArrayVector->sizeAt(decodedRow);
+    auto otherSampleOffset = sampleArrayVector->offsetAt(decodedRow);
+
+    if (!accumulator->initialSamples) {
+      auto initialSampleSize = initialSampleArrayVector->sizeAt(decodedRow);
+      if (initialSampleSize > 0) {
+        auto initialSampleOffset =
+            initialSampleArrayVector->offsetAt(decodedRow);
+        auto initialSampleElements = initialSampleArrayVector->elements();
+
+        accumulator->initialSamples = BaseVector::create(
+            elementType, initialSampleSize, allocator_->pool());
+        accumulator->initialSamples->resize(initialSampleSize);
+        accumulator->initialSamples->copy(
+            initialSampleElements.get(),
+            0,
+            initialSampleOffset,
+            initialSampleSize);
+      }
+    }
+    if (otherSampleSize == 0) {
+      return;
+    }
+
+    auto sampleElements = sampleArrayVector->elements();
+    DecodedVector decodedSampleElements(*sampleElements);
+
+    if (otherProcessedCount < accumulator->maxSampleSize) {
+      for (vector_size_t i = 0; i < otherSampleSize; i++) {
+        accumulator->processedCount++;
+        auto elementIndex = otherSampleOffset + i;
+        accumulator->addValueToReservoir(
```

Comment:
When merging two accumulators this call will end up being very expensive, and would need to be optimized..